hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager/src/main/java/org/apache/hadoop/yarn/server/resourcemanager/scheduler/fair/FSAppAttempt.java

```diff
@@ -1106,6 +1111,97 @@ boolean isStarvedForFairShare() {
     return !Resources.isNone(fairshareStarvation);
   }
 
+  /**
+   * Helper method for {@link #getStarvedResourceRequests()}:
+   * Given a map of visited {@link ResourceRequest}s, it checks if
+   * {@link ResourceRequest} 'rr' has already been visited. The map is updated
+   * to reflect visiting 'rr'.
+   */
+  private static boolean checkAndMarkRRVisited(
+      Map<Priority, List<Resource>> visitedRRs, ResourceRequest rr) {
+    Priority priority = rr.getPriority();
+    Resource capability = rr.getCapability();
+    if (visitedRRs.containsKey(priority)) {
+      List<Resource> rrList = visitedRRs.get(priority);
+      if (rrList.contains(capability)) {
```

Comment:
Yeah, looks like there is indeed a bug here. 

Consider an app asks for one container each on two nodes on the same rack:
- If this code encounters either of these node-local requests, it ignores the other node and the rack requests. Ignoring the other node-local request is undesired. 
- if this code encounters the rack-local request, it ignores the node-local requests. This is desired. 

Maybe, on encountering a node-local request, we should mark the rack and ANY as "visited". What do we do when we encounter rack or ANY first? Let me think more about this.